# 数据挖掘-预测贷款用户是否逾期
## Task1：数据分析（2天）

-------
### 1 &emsp;数据类型的分析

**数据说明：** 

这份数据集是金融数据（非原始数据，已经处理过了），我们要做的是预测贷款用户是否会逾期。

表格中 "status" 是结果标签：0表示未逾期，1表示逾期。

 - **导入宏包**

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

import warnings
warnings.filterwarnings('ignore')

import matplotlib as mpl
mpl.rcParams['font.sans-serif']=[u'SimHei']
mpl.rcParams['axes.unicode_minus']=False

 - 导入数据

In [2]:
data = pd.read_csv('./data/data.csv', encoding='gbk')
data.head()

Unnamed: 0   custid                          trade_no bank_card_no  \
0           5  2791858  20180507115231274000000023057383          卡号1   
1          10   534047  20180507121002192000000023073000          卡号1   
2          12  2849787  20180507125159718000000023114911          卡号1   
3          13  1809708  20180507121358683000000388283484          卡号1   
4          14  2499829  20180507115448545000000388205844          卡号1   

   low_volume_percent  middle_volume_percent  \
0                0.01                   0.99   
1                0.02                   0.94   
2                0.04                   0.96   
3                0.00                   0.96   
4                0.01                   0.99   

   take_amount_in_later_12_month_highest  trans_amount_increase_rate_lately  \
0                                      0                               0.90   
1                                   2000                               1.28   
2                                      0                               1.00   
3                                   2000                               0.13   
4                                      0                               0.46   

   trans_activity_month  trans_activity_day  ...  loans_max_limit  \
0                  0.55               0.313  ...           2900.0   
1                  1.00               0.458  ...           3500.0   
2                  1.00               0.114  ...           1600.0   
3                  0.57               0.777  ...           3200.0   
4                  1.00               0.175  ...           2300.0   

   loans_avg_limit  consfin_credit_limit  consfin_credibility  \
0           1688.0                1200.0                 75.0   
1           1758.0               15100.0                 80.0   
2           1250.0                4200.0                 87.0   
3           1541.0               16300.0                 80.0   
4           1630.0                8300.0                 79.0   

   consfin_org_count_current  consfin_product_count  consfin_max_limit  \
0                        1.0                    2.0             1200.0   
1                        5.0                    6.0            22800.0   
2                        1.0                    1.0             4200.0   
3                        5.0                    5.0            30000.0   
4                        2.0                    2.0             8400.0   

   consfin_avg_limit  latest_query_day  loans_latest_day  
0             1200.0              12.0              18.0  
1             9360.0               4.0               2.0  
2             4200.0               2.0               6.0  
3            12180.0               2.0               4.0  
4             8250.0              22.0             120.0  

[5 rows x 90 columns]

 - 查看数据属性

In [3]:
data.info()

print()
print("共有数据集：", data.shape[0])

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4754 entries, 0 to 4753
Data columns (total 90 columns):
Unnamed: 0                                    4754 non-null int64
custid                                        4754 non-null int64
trade_no                                      4754 non-null object
bank_card_no                                  4754 non-null object
low_volume_percent                            4752 non-null float64
middle_volume_percent                         4752 non-null float64
take_amount_in_later_12_month_highest         4754 non-null int64
trans_amount_increase_rate_lately             4751 non-null float64
trans_activity_month                          4752 non-null float64
trans_activity_day                            4752 non-null float64
transd_mcc                                    4752 non-null float64
trans_days_interval_filter                    4746 non-null float64
trans_days_interval                           4752 non-null float64
regional_mobility

 - 数据类型

In [4]:
for i,name in enumerate(data.columns):
    name_sum = data[name].value_counts().shape[0]
    print("{:2}、{:40}      The number of types of features is：{}".format(i + 1, name, name_sum))

 1、Unnamed: 0                                    The number of types of features is：4754
 2、custid                                        The number of types of features is：4754
 3、trade_no                                      The number of types of features is：4754
 4、bank_card_no                                  The number of types of features is：1
 5、low_volume_percent                            The number of types of features is：40
 6、middle_volume_percent                         The number of types of features is：90
 7、take_amount_in_later_12_month_highest         The number of types of features is：166
 8、trans_amount_increase_rate_lately             The number of types of features is：782
 9、trans_activity_month                          The number of types of features is：84
10、trans_activity_day                            The number of types of features is：512
11、transd_mcc                                    The number of types of features is：41
12、trans_days_interval_filter      

 - 数据统计

In [5]:
data.describe()

Unnamed: 0        custid  low_volume_percent  middle_volume_percent  \
count   4754.000000  4.754000e+03         4752.000000            4752.000000   
mean    6008.414178  1.690993e+06            0.021806               0.901294   
std     3452.071428  1.034235e+06            0.041527               0.144856   
min        5.000000  1.140000e+02            0.000000               0.000000   
25%     3106.000000  7.593358e+05            0.010000               0.880000   
50%     6006.500000  1.634942e+06            0.010000               0.960000   
75%     8999.000000  2.597905e+06            0.020000               0.990000   
max    11992.000000  4.004694e+06            1.000000               1.000000   

       take_amount_in_later_12_month_highest  \
count                            4754.000000   
mean                             1940.197728   
std                              3923.971494   
min                                 0.000000   
25%                                 0.000000   
50%                               500.000000   
75%                              2000.000000   
max                             68000.000000   

       trans_amount_increase_rate_lately  trans_activity_month  \
count                        4751.000000           4752.000000   
mean                           14.160674              0.804411   
std                           694.180473              0.196920   
min                             0.000000              0.120000   
25%                             0.615000              0.670000   
50%                             0.970000              0.860000   
75%                             1.600000              1.000000   
max                         47596.740000              1.000000   

       trans_activity_day   transd_mcc  trans_days_interval_filter  ...  \
count         4752.000000  4752.000000                 4746.000000  ...   
mean             0.365425    17.502946                   29.029920  ...   
std              0.170196     4.475616                   22.722432  ...   
min              0.033000     2.000000                    0.000000  ...   
25%              0.233000    15.000000                   16.000000  ...   
50%              0.350000    17.000000                   23.000000  ...   
75%              0.480000    20.000000                   32.000000  ...   
max              0.941000    42.000000                  285.000000  ...   

       loans_max_limit  loans_avg_limit  consfin_credit_limit  \
count      4457.000000      4457.000000           4457.000000   
mean       3390.038142      1820.357864           9187.009199   
std        1474.206546       583.418291           7371.257043   
min           0.000000         0.000000              0.000000   
25%        2300.000000      1535.000000           4800.000000   
50%        3100.000000      1810.000000           7700.000000   
75%        4300.000000      2100.000000          11700.000000   
max       10000.000000      6900.000000          87100.000000   

       consfin_credibility  consfin_org_count_current  consfin_product_count  \
count          4457.000000                4457.000000            4457.000000   
mean             76.042630                   4.732331               5.227507   
std              14.536819                   2.974596               3.409292   
min               0.000000                   0.000000               0.000000   
25%              77.000000                   2.000000               3.000000   
50%              79.000000                   4.000000               5.000000   
75%              80.000000                   7.000000               7.000000   
max              87.000000                  18.000000              20.000000   

       consfin_max_limit  consfin_avg_limit  latest_query_day  \
count        4457.000000        4457.000000       4450.000000   
mean        16153.690823        8007.696881         24.112809   
std         14301.037628        5679.418585         37.725724   
min             0

 - **结果标签分布情况**

In [6]:
data.status.value_counts()

0    3561
1    1193
Name: status, dtype: int64

**结论一：** 可以看到数据共有4754行，其中浮点型数据70列、整数型13列、object 型7列，**需要对 object 类型的数据进行处理才能建立模型。**

**结论二：** 数据在整体上是相当完整的，**除了 student_feature 这个标签有较多缺失值以外，其他数据仅有很小的缺失。**

**结论三：** **大部分数据特征存在拖尾现象，** 即特征的 max(min) 明显偏离其 mean ，需要对这些数据进行进一步的处理才能建立准确的模型。

**结论四：** **数据集正例与负例数量存在显著差异，** 需要使用 **分层抽样** 的方式对数据集进行抽样划分，否则会导致 **模型对未逾期结果预测较为准确，而对逾期结果预测偏差较大。**

**注意：** 有的列像 'first_transaction_time'，它的值其实是日期的形式类似 20130817 这样，**pandas 把它认为是 int 型的**，其实不是，这也需要注意。

-------
### 2&emsp;无关特征的删除

显然，上述特征中存在一些与贷款用户是否会逾期无关的特征。

经分析，无关特征及字段如下：
 - 'Unnamed: 0'
 - 客户id：'custid'
 - 流水号：'trade_no'
 - 卡号：'bank_card_no'
 - 资源：'source'
 - 客户姓名：'id_name'
 - 最新查询时间：'latest_query_time'
 - 最新贷款时间：'loans_latest_time'

In [7]:
data = data.drop(['Unnamed: 0', 'custid', 'trade_no', 'bank_card_no', 'source', 'id_name', 'latest_query_time','loans_latest_time'], axis = 1)
data.head()

low_volume_percent  middle_volume_percent  \
0                0.01                   0.99   
1                0.02                   0.94   
2                0.04                   0.96   
3                0.00                   0.96   
4                0.01                   0.99   

   take_amount_in_later_12_month_highest  trans_amount_increase_rate_lately  \
0                                      0                               0.90   
1                                   2000                               1.28   
2                                      0                               1.00   
3                                   2000                               0.13   
4                                      0                               0.46   

   trans_activity_month  trans_activity_day  transd_mcc  \
0                  0.55               0.313        17.0   
1                  1.00               0.458        19.0   
2                  1.00               0.114        13.0   
3                  0.57               0.777        22.0   
4                  1.00               0.175        13.0   

   trans_days_interval_filter  trans_days_interval  regional_mobility  ...  \
0                        27.0                 26.0                3.0  ...   
1                        30.0                 14.0                4.0  ...   
2                        68.0                 22.0                1.0  ...   
3                        14.0                  6.0                3.0  ...   
4                        66.0                 42.0                1.0  ...   

   loans_max_limit  loans_avg_limit  consfin_credit_limit  \
0           2900.0           1688.0                1200.0   
1           3500.0           1758.0               15100.0   
2           1600.0           1250.0                4200.0   
3           3200.0           1541.0               16300.0   
4           2300.0           1630.0                8300.0   

   consfin_credibility  consfin_org_count_current  consfin_product_count  \
0                 75.0                        1.0                    2.0   
1                 80.0                        5.0                    6.0   
2                 87.0                        1.0                    1.0   
3                 80.0                        5.0                    5.0   
4                 79.0                        2.0                    2.0   

   consfin_max_limit  consfin_avg_limit  latest_query_day  loans_latest_day  
0             1200.0             1200.0              12.0              18.0  
1            22800.0             9360.0               4.0               2.0  
2             4200.0             4200.0               2.0               6.0  
3            30000.0            12180.0               2.0               4.0  
4             8400.0             8250.0              22.0             120.0  

[5 rows x 82 columns]

--------
### 3&emsp;数据类型转换

经过无关特征删除后，我们还有两个特征需要进行转换：

 -  object 类型特征（城市等级：'reg_preference_for_trad'）
 
 
 - 时间格式特征（首次交易时间：'first_transaction_time_day'）

**3.1&ensp;&ensp;reg_preference_for_trad 转换**

首先我们查看一下该特征下有哪些数据

In [8]:
print(data.reg_preference_for_trad.unique())

['一线城市' '三线城市' '境外' '二线城市' '其他城市' nan]


这里我们先简单的进行数据转换，特征提取时可以采用 one-hot 编码改进进行改进。

In [9]:
dic = {'一线城市': 1,
       '二线城市': 2,
       '三线城市': 3,
       '其他城市': 4,
       '境外': 0}
data.reg_preference_for_trad = data.reg_preference_for_trad.map(dic)

**3.2  first_transaction_time_day 转换**

对于时间数据，这里将年月日分别建立新的列处理

In [10]:
tmpdf = pd.DataFrame()
tmpdf['first_transaction_time_year'] = pd.to_datetime(data['first_transaction_time'], format='%Y%m%d').dt.year
tmpdf['first_transaction_time_month'] = pd.to_datetime(data['first_transaction_time'], format='%Y%m%d').dt.month
tmpdf['first_transaction_time_day'] = pd.to_datetime(data['first_transaction_time'], format='%Y%m%d').dt.day
data[tmpdf.columns] = tmpdf
data = data.drop('first_transaction_time_day', axis = 1)

**3.3&ensp;&ensp;结果查看**

In [11]:
data = data.convert_objects(convert_numeric=True)
print(data.dtypes.value_counts())

float64    73
int64      11
dtype: int64


至此，数据类型的转换工作完成

--------
### 4&emsp;处理缺失值

 - 删除缺失值较多的行列
 
 若某个特征的缺失值超过 30%，那么它会损失特征的关键信息。
 
 同样若某个样本的缺失值过多，那么它也不再具有统计学意义。
 
 因此我们对这样的行或列直接进行删除。

In [12]:
data = data.dropna(axis=1, thresh = 1000)#删除缺失值超过 70% 的列
data = data.dropna(axis=0, thresh = 75)#删除缺失值超过7个的行

 - 填充缺失值
 
 对于剩余的缺失值，这里采用 **平均值填充法** 进行填充。

In [13]:
data = data.fillna(data.median())

data

low_volume_percent  middle_volume_percent  \
0                   0.01                   0.99   
1                   0.02                   0.94   
2                   0.04                   0.96   
3                   0.00                   0.96   
4                   0.01                   0.99   
5                   0.02                   0.98   
6                   0.02                   0.98   
8                   0.03                   0.65   
9                   0.01                   0.99   
10                  0.01                   0.99   
11                  0.01                   0.99   
12                  0.00                   0.73   
13                  0.01                   0.79   
14                  0.01                   0.89   
15                  0.01                   0.99   
16                  0.02                   0.98   
17                  0.02                   0.98   
18                  0.01                   0.99   
19                  0.01                   0.96   
20                  0.02                   0.98   
21                  0.02                   0.98   
22                  0.00                   0.59   
23                  0.01                   0.99   
24                  0.02                   0.78   
25                  0.40                   0.60   
26                  0.02                   0.96   
27                  0.01                   0.91   
28                  0.01                   0.99   
30                  0.02                   0.93   
31                  0.02                   0.98   
...                  ...                    ...   
4723                0.01                   0.96   
4724                0.02                   0.98   
4725                0.01                   0.99   
4726                0.00                   1.00   
4727                0.02                   0.98   
4728                0.06                   0.94   
4729                0.17                   0.83   
4730                0.01                   0.99   
4731                0.00                   0.58   
4732                0.11                   0.89   
4733                0.03                   0.97   
4734                0.00                   1.00   
4735                0.01                   0.61   
4736                0.05                   0.95   
4737                0.01                   0.99   
4738                0.02                   0.98   
4739                0.05                   0.62   
4740                0.07                   0.93   
4741                0.00                   0.83   
4742                0.02                   0.98   
4743                0.00                   1.00   
4744                0.01                   0.96   
4745                0.04                   0.96   
4747                0.02                   0.40   
4748                0.09                   0.91   
4749                0.05                   0.68   
4750                0.01                   0.99   
4751                0.00                   1.00   
4752                0.05                   0.95   
4753                0.01                   0.99   

      take_amount_in_later_12_month_highest  \
0                                         0   
1                                      2000   
2                                         0   
3                                      2000   
4                                         0   
5                                      2000   
6                                         0   
8                                         0   
9                                       500   
10                                        0   
11                                      500   
12                                     5000   
13                                        0   
14                                     4000   
15                                        0   
16                                        0   
17                                      300 

-------
### 5&emsp;异常值检验

这里我们采用 LOF 算法进行异常值检验。

LOF 算法参考文献：[LOF离群因子检测算法](https://zhuanlan.zhihu.com/p/37753692)

 - **构造检验函数**

In [14]:
def localoutlierfactor(data, predict, k):
    from sklearn.neighbors import LocalOutlierFactor
    LOF = LocalOutlierFactor(n_neighbors=k + 1, algorithm='auto', contamination=0.1, n_jobs=-1)
    LOF.fit(data)
    # 记录 k 邻域距离
    predict['k distances'] = LOF.kneighbors(predict)[0].max(axis=1)
    # 记录 LOF 离群因子，做相反数处理
    predict['local outlier factor'] = -LOF._decision_function(predict.iloc[:, :-1])
    return predict

def lof(data, predict=None, k=10, method=1):
    import pandas as pd
    # 判断是否传入测试数据，若没有传入则测试数据赋值为训练数据
    try:
        if predict == None:
            predict = data.copy()
    except Exception:
        pass
    predict = pd.DataFrame(predict)
    # 计算 LOF 离群因子
    predict = localoutlierfactor(data, predict, k)
    # 根据阈值划分离群点与正常点
    outliers = predict[predict['local outlier factor'] > method].sort_values(by='local outlier factor')
    inliers = predict[predict['local outlier factor'] <= method].sort_values(by='local outlier factor')
    return outliers, inliers

 - **计算并删除异常样本**

In [15]:
out_data, in_data = lof(data, k=10, method = 2)
data.drop(out_data.index, axis = 0)

out_data

low_volume_percent  middle_volume_percent  \
1002                0.00                   0.71   
1576                0.00                   0.76   
2938                0.01                   0.73   
2603                0.00                   0.28   
471                 0.01                   0.88   

      take_amount_in_later_12_month_highest  \
1002                                   2000   
1576                                   3300   
2938                                   1000   
2603                                   4000   
471                                       0   

      trans_amount_increase_rate_lately  trans_activity_month  \
1002                               0.55                  1.00   
1576                               0.02                  0.98   
2938                               0.36                  1.00   
2603                              13.70                  1.00   
471                                0.45                  0.78   

      trans_activity_day  transd_mcc  trans_days_interval_filter  \
1002               0.683        15.0                        15.0   
1576               0.941        17.0                        24.0   
2938               0.897        19.0                        21.0   
2603               0.886        25.0                         7.0   
471                0.452        18.0                        30.0   

      trans_days_interval  regional_mobility  ...  consfin_org_count_current  \
1002                 10.0                2.0  ...                       10.0   
1576                  5.0                2.0  ...                        2.0   
2938                  5.0                1.0  ...                        3.0   
2603                  5.0                2.0  ...                        4.0   
471                  23.0                2.0  ...                        6.0   

      consfin_product_count  consfin_max_limit  consfin_avg_limit  \
1002                   11.0            25200.0            10840.0   
1576                    2.0            12000.0            11850.0   
2938                    3.0            13200.0             5733.0   
2603                    5.0            17400.0             9000.0   
471                     7.0            37200.0            14016.0   

      latest_query_day  loans_latest_day  first_transaction_time_year  \
1002              11.0               8.0                       2016.0   
1576               3.0              15.0                       2014.0   
2938              15.0              23.0                       2016.0   
2603              23.0              -1.0                       2015.0   
471               21.0               5.0                       2015.0   

      first_transaction_time_month   k distances  local outlier factor  
1002                          10.0  4.139863e+05              2.086765  
1576                           9.0  1.002926e+06              2.908565  
2938                           1.0  2.383146e+06              6.386036  
2603                           5.0  1.305017e+07              7.071480  
471                            7.0  1.440188e+06              7.183463  

[5 rows x 86 columns]

-------
### 6&emsp;数据集切分

 - **分层抽样重构数据集**

In [16]:
dfstatus0 = data[data.status == 0]
dfstatus1 = data[data.status == 1]
dfstatus1.sample(frac=data.status.value_counts()[1], replace=True, random_state=2018)

newdata = pd.concat([dfstatus0, dfstatus1], ignore_index=False)
newdata.sort_index(inplace=True)
newdata = newdata.reset_index(drop=True)
newdata.head()

low_volume_percent  middle_volume_percent  \
0                0.01                   0.99   
1                0.02                   0.94   
2                0.04                   0.96   
3                0.00                   0.96   
4                0.01                   0.99   

   take_amount_in_later_12_month_highest  trans_amount_increase_rate_lately  \
0                                      0                               0.90   
1                                   2000                               1.28   
2                                      0                               1.00   
3                                   2000                               0.13   
4                                      0                               0.46   

   trans_activity_month  trans_activity_day  transd_mcc  \
0                  0.55               0.313        17.0   
1                  1.00               0.458        19.0   
2                  1.00               0.114        13.0   
3                  0.57               0.777        22.0   
4                  1.00               0.175        13.0   

   trans_days_interval_filter  trans_days_interval  regional_mobility  ...  \
0                        27.0                 26.0                3.0  ...   
1                        30.0                 14.0                4.0  ...   
2                        68.0                 22.0                1.0  ...   
3                        14.0                  6.0                3.0  ...   
4                        66.0                 42.0                1.0  ...   

   consfin_credit_limit  consfin_credibility  consfin_org_count_current  \
0                1200.0                 75.0                        1.0   
1               15100.0                 80.0                        5.0   
2                4200.0                 87.0                        1.0   
3               16300.0                 80.0                        5.0   
4                8300.0                 79.0                        2.0   

   consfin_product_count  consfin_max_limit  consfin_avg_limit  \
0                    2.0             1200.0             1200.0   
1                    6.0            22800.0             9360.0   
2                    1.0             4200.0             4200.0   
3                    5.0            30000.0            12180.0   
4                    2.0             8400.0             8250.0   

   latest_query_day  loans_latest_day  first_transaction_time_year  \
0              12.0              18.0                       2013.0   
1               4.0               2.0                       2016.0   
2               2.0               6.0                       2017.0   
3               2.0               4.0                       2013.0   
4              22.0             120.0                       2017.0   

   first_transaction_time_month  
0                           8.0  
1                           4.0  
2                           6.0  
3                           5.0  
4                           3.0  

[5 rows x 84 columns]

 - **数据集切分**

In [17]:
from sklearn.model_selection import train_test_split

train_data, test_data = train_test_split(newdata, test_size=0.3, random_state=2018)
train_data.to_csv('./data/train_data.csv', index=False, header=True)
test_data.to_csv('./data/test_data.csv', index=False, header=True)
newdata.to_csv('./data/newdata.csv', index=False, header=True)